frame/support/src/executor.rs

```diff
@@ -0,0 +1,620 @@
+	/// under any circumstance. Consuming *less* than `max_weight` is allowed.
+	///
+	/// A tuple is returned, where the items are as follows:
+	///   1. Option<Self>, where `None` means that this task is now complete (and shall not be kept
+	///      in storage anymore), and `Some(_)` indicating that this task is not yet complete, and
+	///      should be executed at a later time.
+	///   2. The actual amount of weight that was consumed. Must always be less than `max_weight`.
+	///      parameter.
+	///
+	/// It is critically important for a task to only return a non-zero consumed weight **ONLY if it
+	/// _actually did something_**. If a positive weight is returned, then an executor could
+	/// interpret this as a task that could use another execution slot, and continue the execution
+	/// potentially for numerous iterations.
+	fn execute(self, max_weight: Weight) -> (Option<Self>, Weight);
+
+	/// The leftover weight that this task expects to execute, if any.
+	#[cfg(test)]
+	fn leftover(&self) -> Weight;
+}
+
+#[cfg(any(test, feature = "std"))]
+impl RuntimeTask for () {
+	fn execute(self, _: Weight) -> (Option<Self>, Weight) {
+		(None, 0)
+	}
+	#[cfg(test)]
+	fn leftover(&self) -> Weight {
+		0
+	}
+}
+
+/// Common trait for a executor that is stored as a storage item.
+pub trait StoredExecutor {
+	/// The task type used by this executor.
+	type Task: RuntimeTask;
+
+	/// Something that can define how much weight quote this executor is allowed to use per
+	/// execution.
+	type Quota: Get<Weight>;
+
+	/// Execute all tasks based on an unspecified strategy, consuming at most `Self::Quota` and
+	/// returning the actual amount of weight consumed.
+	///
+	/// The returned weight must take into account the cost of internal operations of the
+	/// implementation, such as scheduling, as well. But, it DOES NOT take into account any
+	/// potential storage operations that needed to be performed to fetch `Self` from storage.
+	///
+	/// A  sensible patter of using an implementation of this trait is therefore:
+	///
+	/// ```ignore
+	/// let mut consumed = <StorageItemExecutor<T>>::mutate(|e| e.execute());
+	/// consumed += <weight of 1 read and write for the mutate call above>;
+	/// ```
+	// TODO: while the work that this does itself is pretty negligible, must benchmark it anyhow and
+	// take it into account.
+	fn execute(&mut self) -> Weight;
+
+	/// Create a new (empty) instance of [`Self`].
+	fn new() -> Self;
+
+	/// Add a new task to the internal state.
+	fn add_task(&mut self, task: Self::Task);
+
+	/// Remove all tasks, without executing any of them.
+	fn clear(&mut self);
+
+	/// Remove a single task.
+	fn remove(&mut self, task: Self::Task);
+
+	/// Returns the number of current tasks.
+	fn count(&self) -> usize;
+
+	/// Return a vector of all tasks.
+	#[cfg(any(test, feature = "std"))]
+	fn tasks(&self) -> Vec<Self::Task>;
+	// TODO: providing an iter() might also be good.
```

Comment:
I would really like to have those as a github issue rather than a TODO comment. But this is a personal preference. Feel free to ignore.

Reply:
A PR shall never be merged with TODOs, these are notes that I want my reviewers to also see and be aware that I maybe intend to do them.